androguard/core/apk/__init__.py

```diff
@@ -1499,21 +1501,153 @@ def is_androidtv(self) -> bool:
 
         :param filename: Signature filename in APK
         :returns: DER coded X.509 certificate as binary
         """
+
+        # Get the signature
         pkcs7message = self.get_file(filename)
+        # Get the .SF
+        sf_filename = os.path.splitext(filename)[0] + '.SF'
+        sf_object = self.get_file(sf_filename)
+        # Load the signature
+        signed_data = cms.ContentInfo.load(pkcs7message)
+        # Locate the SignerInfo structure
+        signer_infos = signed_data['content']['signer_infos']
+        if not signer_infos:
+            raise ValueError('No signer information found in the PKCS7 object. The APK may not be properly signed.')
+
+        # Prior to Android N, Android attempts to verify only the first SignerInfo. From N onwards, Android attempts
+        # to verify all SignerInfos and then picks the first verified SignerInfo.
+        min_sdk_version = self.get_min_sdk_version()
+        if min_sdk_version is None or int(min_sdk_version) < 24:  # AndroidSdkVersion.N
+            logger.info(f"minsdkversion: {min_sdk_version} is less than 24. Getting the first signerInfo only!")
+            unverified_signer_infos_to_try = [signer_infos[0]]
+        else:
+            unverified_signer_infos_to_try = signer_infos
+
+        # Extract certificates from the PKCS7 object
+        certificates = signed_data['content']['certificates']
+        matching_certificate_verified = None
+        for signer_info in unverified_signer_infos_to_try:
+            matching_certificate_verified = self.verify_signer_info_against_sig_file(
+                signed_data,
+                certificates,
+                signer_info,
+                sf_object)
+            if matching_certificate_verified is not None:
+                # Return the first certificate that was verified
+                break
+        if not matching_certificate_verified:
+            logger.warning(f"minSdkVersion: {min_sdk_version}, # of SignerInfos: {len(unverified_signer_infos_to_try)}. None Verified!")
+        return matching_certificate_verified
+
+    def verify_signer_info_against_sig_file(self, signed_data, certificates, signer_info, sf_object):
+        matching_certificate = self.find_certificate(certificates, signer_info)
+        matching_certificate_verified = None
+        if matching_certificate is None:
+            logger.info("Signing certificate referenced in SignerInfo not found in SignedData")
+        else:
+            if signer_info['signed_attrs'].native:
+                logger.info("Signed Attributes detected!")
+                signed_attrs = signer_info['signed_attrs']
+                signed_attrs_dict = OrderedDict()
+                for attr in signed_attrs:
+                    if attr['type'].dotted in signed_attrs_dict:
+                        logger.error(f"Duplicate signed attribute: {attr['type'].dotted}")
+                        return None
```

Comment:
Would it be more graceful to instead return None and let the use handle this however they wish?
Since there might be cases where one would wish to analyze an APK that indeed fails the verification.

Reply:
Yes? I would say `get_certificate_der()` should return `None` as verification failed and there is thus no correct cert for it to return. But currently, it is `verify_signer_info_against_sig_file()` that simply returns `None`, and `get_certificate_der()` will return the wrong certificate if another call to `verify_signer_info_against_sig_file()` does return a cert.

Reply:
There are three places in the code which currently raise a `ValueError` instead of causing `get_certificate_der()` to return `None`. I would say the same logic applies there.